### PR TITLE
Indexed "Getters", Xpath, Isolates

### PR DESCRIPTION
**Pull request by blakebyrnes**
This Pr does a few things:

    Fix a bug recovering the db state so that abstract is restored in the db
    Enable "getter" access via webidls by adding a proxy in objects that support getters and delegating to the named version of the function
    Fix typing of the StateMachine
    Enable xpath access

    to do so, we needed to change
    XPathResult is a "NodeAttached" type of element - so it can be "locked-in" by awaiting it

    Switch the AwaitedHandler to get a delegate pattern so that it doesn't explode if it's not initialized before any classes in awaiteddom are accessed
    Change all parameters that are sent into methods to be "Isolates" (as applicable).

### CODE REVIEW

File: src/TsStateMachinePrinter.ts [hunk 1]
```diff
@@ -16,18 +16,32 @@ export default class TsStateMachinePrinter {
...
-  constructor(i: Types.Interface, printer: Printer, components: Components, options: IOptions) {
+  constructor(
+    i: Types.Interface,
+    printer: Printer,
+    components: Components,
+    options: IOptions,
+    iterable?: {
+      hasIterable: () => boolean;
+      hasMaplikeSequence: boolean;
+    },
```
Comment (calebjclark) [inline src/TsStateMachinePrinter.ts:32]: I find this really confusing, specifically how the other classes are passing "iterable" in (this.bodyPrinter.iteratorExtractor). Why not pass these in as options?

File: src/TsBuildMixin.ts [hunk 2]
```diff
@@ -87,7 +93,13 @@ export default class TsBuildMixin {
     const handlerClassName = `${isAwaited ? 'Awaited' : 'Detached'}Handler`;
     const printer = new Printer();
     const stateOptions = { domType: this.domType, buildType: BuildType.base };
-    new TsStateMachinePrinter(this.i, printer, this.components, stateOptions).printInitializer();
+    new TsStateMachinePrinter(
+      this.i,
+      printer,
+      this.components,
+      stateOptions,
+      this.bodyPrinter.iteratorExtractor,
```
Comment (calebjclark) [inline src/TsBuildMixin.ts:101]: See my comment in TsStateMachinePrinter.ts below.

File: src/exports-for-base/StateMachine.ts [hunk 3]
```diff
@@ -1,20 +1,28 @@
 const globalMap = new WeakMap();
+const proxyMap = new WeakMap();
```
Comment (calebjclark) [inline src/exports-for-base/StateMachine.ts:2]: What is the use case for proxyMap? I see that it's used for indexers, but why is this needed in addition to globalMap?

File: src/TsStateMachinePrinter.ts (hunk 1 above)
Comment (blakebyrnes) [reply src/TsStateMachinePrinter.ts:32]: It felt wrong to pass the whole IteratorExtractor into the state machine printer, so I just defined the properties I wanted available to this class. I could just pass the parameters individually too (I had that initially), but it seemed more bug prone than to leave them on the source object so they all change in unison if some other extractor messed with the underlying properties

File: src/exports-for-base/StateMachine.ts (hunk 3 above)
Comment (blakebyrnes) [reply src/exports-for-base/StateMachine.ts:2]: When you use awaitedhandler in the classes, Internal access to "this" was not finding itself in the globalmap. But if you get state for the object from "outside" the class, it seems to be using the proxy object "identity". To be honest, I don't quite understand the behavior I was seeing here, but it's likely because the constructor returns a Proxy to the outside world.

File: src/TsStateMachinePrinter.ts (hunk 1 above)
Comment (calebjclark) [reply src/TsStateMachinePrinter.ts:32]: Hrm... k. I find it much easier to follow the code when it's clear where these iterable methods are coming from... especially since all the this.bodyPrinter.iteratorExtractor is what's being passed in from all the class initializers.

Comment (blakebyrnes) [reply src/TsStateMachinePrinter.ts:32]: I'm fine with changing it. Are you saying we should pass in individual variables though, or the IteratorExtractor?

Comment (calebjclark) [reply src/TsStateMachinePrinter.ts:32]: You make a good point about passing them in individually seems more error prone... I would recommend converting `iterable` to `iterableExtractor: IteratorExtractor`.

Comment (calebjclark) [reply src/TsStateMachinePrinter.ts]: That makes it very clear within TsStateMachinePrinter.ts what I'm looking at, and it minimizes any `this` issues from/if destructuring.

Comment (blakebyrnes) [reply src/TsStateMachinePrinter.ts]: Ok, I changed it
